lib/SPIRV/SPIRVLowerBitCastToNonStandardType.cpp

```diff
@@ -0,0 +1,200 @@
+
+#include <utility>
+
+using namespace llvm;
+
+namespace SPIRV {
+
+static VectorType *getVectorType(Type *Ty) {
+  if (!Ty)
+    return nullptr;
+  if (auto *ElemTy = dyn_cast<PointerType>(Ty))
+    Ty = ElemTy->getElementType();
+  return dyn_cast<VectorType>(Ty);
+}
+
+/// Since SPIR-V does not support non-standard vector types, instructions using
+/// these types should be replaced in a special way to avoid using of
+/// unsupported types.
+/// lowerBitCastToNonStdVec function is designed to avoid using of bitcast to
+/// unsupported vector types instructions and should be called if similar
+/// instructions have been encountered in input LLVM IR.
+bool lowerBitCastToNonStdVec(Instruction *Inst,
+                             std::vector<Instruction *> &InstsToErase) {
+  Instruction *OldInstIter(Inst);
+  IRBuilder<> Builder(OldInstIter);
+  Value *NewValue = OldInstIter->getOperand(0);
+  VectorType *NewVecTy = getVectorType(NewValue->getType());
+  InstsToErase.push_back(OldInstIter);
+  // Handle addrspacecast instruction after bitcast if present
+  for (auto *U : OldInstIter->users()) {
+    if (auto *ASCastInst = dyn_cast<AddrSpaceCastInst>(U)) {
+      unsigned DestAS = ASCastInst->getDestAddressSpace();
+      auto *NewVecPtrTy = NewVecTy->getPointerTo(DestAS);
+      // AddrSpaceCast is created explicitly instead of using method
+      // IRBuilder<>.CreateAddrSpaceCast because IRBuilder doesn't create
+      // separate instruction for constant values. Whereas SPIR-V translator
+      // doesn't like several nested instructions in one.
+      NewValue = new AddrSpaceCastInst(NewValue, NewVecPtrTy);
+      Builder.Insert(NewValue);
+      OldInstIter = ASCastInst;
+      InstsToErase.push_back(OldInstIter);
+      break;
+    }
+  }
+  // Handle load instruction which is following the bitcast in the pattern
+  for (auto *U : OldInstIter->users()) {
+    if (auto *LI = dyn_cast<LoadInst>(U)) {
+      NewValue = Builder.CreateLoad(NewVecTy, NewValue);
+      OldInstIter = LI;
+      InstsToErase.push_back(OldInstIter);
+      break;
+    }
+  }
+  // Handle extractelement instruction which is following the load
+  for (auto *U : OldInstIter->users()) {
+    if (auto *EEI = dyn_cast<ExtractElementInst>(U)) {
+      VectorType *OldVecTy = getVectorType(Inst->getType());
+      uint64_t NumElemsInOldVec = OldVecTy->getElementCount().getValue();
+      uint64_t NumElemsInNewVec = NewVecTy->getElementCount().getValue();
+      uint64_t ElemIdx =
+          cast<ConstantInt>(EEI->getIndexOperand())->getSExtValue() /
+          (NumElemsInOldVec / NumElemsInNewVec);
+      NewValue = Builder.CreateExtractElement(NewValue, ElemIdx);
+      NewValue = Builder.CreateTrunc(NewValue, OldVecTy->getElementType());
+      OldInstIter = EEI;
+      InstsToErase.push_back(OldInstIter);
+      break;
+    }
+  }
+  OldInstIter->replaceAllUsesWith(NewValue);
+  return true;
+}
+
+class SPIRVLowerBitCastToNonStandardTypePass {
```

Comment:
If I understand correctly, to make it a proper pass you must [inherit from `PassInfoMixin`](https://llvm.org/docs/WritingAnLLVMNewPMPass.html#basic-code-required)